### PR TITLE
Refactor registry and factories 

### LINKED ISSUE #24218: Rework Registry utilizing polymorphism on RegistryEntry
## Reason
I need to obtain the registered class name from the Registry, given a typename template parameter. We cannot currently do that before creating the object in the factory (but we need the class name as a string to do that). Demangle does not work for classes that are typedefs of templates (everything is_ad for example). This will get rid of the unintelligible function pointers and replace them by virtual proxy functions.

## Design
Create `RegistryEntryBase` and `RegistryEntry<T>`, add virtual proxies, get rid of function pointers.

## Impact
Better code, added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #24218

Let's see if this works.

### CODE REVIEW

**Comment by dschwen** [comment]
Forgot one more virtual destructor...

**Comment by lindsayad** [comment]
Can you convert this to draft while you are battling with CIVET?

**Comment by dschwen** [comment]
Ok, this fixes the object registration tests. Next the multi app tests need fixing.

**Comment by dschwen** [comment]
Ok, this is ready for some input. I can fix the apps. The PR also removes a bunch of deprecated registration stuff.

**Review by loganharbour — CHANGES_REQUESTED** [review]
Looks great to me. There's some (in my opinion) unnecessary exposure of smart pointers, but that's really my only concern. Beautiful cleanup

File: framework/include/base/AppFactory.h
Comment (dschwen) [reply framework/include/base/AppFactory.h:101]: It replaces `registeredObjectsBegin` and `registeredObjectsEnd`, but I suppose there is no reason not to make it const

File: framework/src/base/MooseApp.C
Comment (dschwen) [reply framework/src/base/MooseApp.C:634]: Ugh, that was a failed attempt (not just because I mistyped empty :-D)

File: framework/include/actions/ActionFactory.h
Comment (dschwen) [reply framework/include/actions/ActionFactory.h:70]: That'd be a pain in the butt. The objects are co-owned between the Registry and the Action Factory. We could do naked pointers, since the teardown is broken in clang and we need to rely on the OS to free the memory when the application exits. IMO this is fine, as there are no other resources to be freed and we need the registry and factory until the program quits.

File: framework/include/base/Registry.h
Comment (dschwen) [reply framework/include/base/Registry.h:278]: No, it cannot, because both methods will be compiled, but only one may be called.

File: framework/include/base/Registry.h
Comment (dschwen) [reply framework/include/base/Registry.h:288]: see above

**Comment by dschwen** [comment]
In the spirit of the recent discussion of commit inflation I have squashed the `More cleanup (#24218)` commit. Take that @lindsayad ! :-D

File: framework/include/actions/ActionFactory.h
Comment (loganharbour) [reply framework/include/actions/ActionFactory.h:70]: Sounds good. thanks for the explanation; just making sure

**Review by loganharbour — COMMENTED** [review]
All good with me. I'll wait for the approval until the app fixes are ready. Does this satisfy your previous problem fully?

**Comment by dschwen** [comment]
> Does this satisfy your previous problem fully?
Yes

**Review by loganharbour — APPROVED** [review]

**Pull request merged**
